### DISCUSSION TITLE
Cannot pull docker image from Github Package Registry by digest

### DISCUSSION THREAD

**Discussion by michaelbeaumont** [community/community#25632] [Actions]
<p>First build, tag and push an image:</p>
<pre><code>$ docker push docker.pkg.github.com/org/repo/image:latest
The push refers to repository [docker.pkg.github.com/org/repo/image]
056157488f63: Layer already exists
556bd7954f59: Layer already exists
bc8ba74a019b: Layer already exists
9b33b8316bd2: Layer already exists
c99984bb0f35: Layer already exists
99add98f81ea: Layer already exists
17cf356f41cc: Layer already exists
4fd9fad7a009: Layer already exists
c9581514d6d0: Layer already exists
46ed3d879948: Layer already exists
fbeeb71995b3: Layer already exists
bb9e1c111e49: Layer already exists
ac3ac7a153b5: Layer already exists
3bfeb766f97b: Layer already exists
ea1227feeccb: Layer already exists
9cae1895156d: Layer already exists
52dba9daa22c: Layer already exists
78c1b9419976: Layer already exists
latest: digest: sha256:2571e10a9946752074aa996358f1cded5bfd707f5517c17317a86e2986eb69c7 size: 4100
</code></pre>
<p>then pulling the same image by digest directly:</p>
<pre><code>$ docker pull docker.pkg.github.com/org/repo/image:latest@sha256:2571e10a9946752074aa996358f1cded5bfd707f5517c17317a86e2986eb69c7
Error response from daemon: manifest for docker.pkg.github.com/org/repo/image@sha256:2571e10a9946752074aa996358f1cded5bfd707f5517c17317a86e2986eb69c7 not found: manifest unknown: Docker image reference image:sha256:2571e10a9946752074aa996358f1cded5bfd707f5517c17317a86e2986eb69c7 not found under repo "org/repo"
</code></pre>
<p>this works with <a href="http://docker.io" rel="nofollow noopener">docker.io</a> as well as other private registries.</p>

**Comment by AndreaGriffiths11** [comment]
<p>HI <a class="mention" href="/michaelbeaumont">@michaelbeaumont</a>,</p>
<p>Thank you for being here! We don’t support pulling an image by digest currently, only by tag.</p>
<p>Here is a link to our docs:</p>
<p><a href="https://help.github.com/en/articles/configuring-docker-for-use-with-github-package-registry#installing-a-package" rel="nofollow noopener">https://help.github.com/en/articles/configuring-docker-for-use-with-github-package-registry#installing-a-package</a></p>

**Comment by edgarnansen** [comment]
<blockquote>
<p>We don’t support pulling an image by digest currently, only by tag.</p>
</blockquote>
<p>Currently means there’s plans to support it in the future? Do you have any updates?<br>
Thanks</p>

**Comment by IGassmann** [comment]
<p>It would be great to be notified of when that would be available. Due to this, Google Cloud Run is currently not compatible with GitHub Packages. That forces us to use Google’s container registry.</p>

**Comment by DrPyser** [comment]
<p>Hi!</p>
<p>Does this mean that github packages’ docker registry cannot be used with docker swarm? When I attempt to (re)-deploy our docker swarm services, which use images hosted on our package registry, I get this error:</p>
<pre><code class="lang-auto">
image docker.pkg.github.com/konversion/myrepo/myimage:mytag could not be accessed on a registry to record  
its digest. Each node will access docker.pkg.github.com/konversion/myrepo/myimage:mytag independently,  
possibly leading to different nodes running different  
versions of the image.

</code></pre>

**Comment by jdavisp3** [comment]
<p>I think this would be a very helpful feature to support. Pulling an image by digest allows you to be very precise about the specific deployment you want. The lack of this feature also means the github registry isn’t compatible with a number of existing tools that assume pulling by digest is possible.</p>

**Comment by splitice** [comment]
<p>I beleive this is the underlying cause of incompatibilities such as <a href="https://github.com/portainer/portainer/issues/3192" rel="nofollow noopener">https://github.com/portainer/portainer/issues/3192</a></p>

**Comment by dfreilich** [comment]
<p>Adding on - this makes the Github Package Registry noncompliant with the <a href="https://github.com/opencontainers/distribution-spec/blob/7229834a8c3c490c2c1b0a30a80049c04abbc08c/spec.md#pulling-an-image-manifest" rel="nofollow noopener">OCI Distribution spec</a>, and is making buildpacks/pack unable to support publishing to GPR, as mentioned <a href="https://github.com/buildpacks/lifecycle/issues/323" rel="nofollow noopener">here</a>.</p>
<p>Are there any updates on a timeline for this?</p>

**Comment by mths0x5f** [comment]
<p>This makes the GitHub Docker Registry almost useless.</p>

**Comment by clarkbw** [comment, marked as answer]
<p>:wave: As noted here in the <a href="https://github.com/containerd/containerd/issues/3291#issuecomment-640253502">containerd issue</a> our Docker service doesn’t currently support pulling by digest.  However we have a new service under private beta which does support pull by digest.  The new service also supports the OCI spec which allows for buildpacks etc.  Please reach out if you’d like to participate.  We have been adding users into the private beta over the past several weeks.</p>

**Comment by freddieptf** [comment]
<p>Hi clarkbw, where do i sign up?</p>

**Comment by clarkbw** [comment]
<p>Email me.  The containerd issue has the required info.</p>

**Comment by iMerica** [comment]
<p>So basically if you deploy with Kubernetes and use digests in logic, you cannot use Github’s Registry. Am I understanding this correctly?</p>

**Comment by jcansdale** [comment]
<p>Hi <a class="mention" href="/imerica">@iMerica</a>,</p>
<p>Yes, you’re understanding correctly.</p>
<p>Could you see the comments at the end of this issue and request access to the private Bata?<br>
</p><aside class="onebox githubissue">
  <header class="source">
      <a href="https://github.com/containerd/containerd/issues/3291" target="_blank" rel="nofollow noopener">github.com/containerd/containerd</a>
  </header>
  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Issue">
	  <svg width="60" height="60" class="github-icon" viewbox="0 0 14 16" aria-hidden="true"><path d="M7 2.3c3.14 0 5.7 2.56 5.7 5.7s-2.56 5.7-5.7 5.7A5.71 5.71 0 0 1 1.3 8c0-3.14 2.56-5.7 5.7-5.7zM7 1C3.14 1 0 4.14 0 8s3.14 7 7 7 7-3.14 7-7-3.14-7-7-7zm1 3H6v5h2V4zm0 6H6v2h2v-2z"></path></svg>
  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/containerd/containerd/issues/3291" target="_blank" rel="nofollow noopener">containerd can't pull image from Github Docker Package Registry</a>
    </h4>

    <div class="github-info">
      <div class="date">
        opened <span class="discourse-local-date" data-format="ll" data-date="2019-05-20" data-time="13:03:34" data-timezone="UTC">01:03PM - 20 May 19 UTC</span>
      </div>


      <div class="user">
        <a href="https://github.com/csantanapr" target="_blank" rel="nofollow noopener">
          <img alt="csantanapr" src="https://user-images.githubusercontent.com/11719160/181092284-356e9ffa-88ec-4e7d-a4d2-591cb074a9fa.jpeg" class="onebox-avatar-inline" width="20" height="20">
          csantanapr
        </a>
      </div>
    </div>
  </div>
</div>

<div class="github-row">
  <p class="github-content">Using the new github docker registry containerd kubernetes can't pull image but using docker engine based k8s works fine.
Steps to reproduce...</p>
</div>

<div class="labels">
    <span style="display:inline-block;margin-top:2px;background-color: #B8B8B8;padding: 2px;border-radius: 4px;color: #fff;margin-left: 3px;">kind/external</span>
</div>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by joshbouganim** [comment]
<p><a class="mention" href="/clarkbw">@clarkbw</a> <a class="mention" href="/imerica">@iMerica</a> , can you confirm that this new service is active in GitHub Enterprise Server 3.0.4?<br>
I am seeing a similar issue</p>
<pre><code class="lang-auto">administrator@gl-mcu-buildserver:~$ docker pull github.company.com/mcu/jarvis-ui/docker-build-cache:latest
WARNING: ⚠️ Failed to pull manifest by the resolved digest. This registry does not
        appear to conform to the distribution registry specification; falling back to
        pull by tag.  This fallback is DEPRECATED, and will be removed in a future
        release.  Please contact admins of https://github.company.com. ⚠️

latest: Pulling from mcu/jarvis-ui/docker-build-cache
540db60ca938: Already exists 
d037ddac5dde: Pulling fs layer 
629719f9106a: Pulling fs layer 
f9ef3a05a91e: Pulling fs layer 
a953f4282204: Waiting 
d4414f840203: Waiting 
b5b7e23f81f8: Waiting 
bc82443eb401: Waiting 
d4bb2cbb6a73: Waiting 
466fe4ff80b1: Waiting 
error pulling image configuration: Get https://storage-mcu.company.com:9000/actions/2225/42e5a880-007a-11ec-9266-87ef7f085566?X-Amz-Algorithm=AWS4-HMAC-SHA256&amp;X-Amz-Credential=Service.MCU%2F20210819%2Fus-east-1%2Fs3%2Faws4_request&amp;X-Amz-Date=20210819T000714Z&amp;X-Amz-Expires=300&amp;X-Amz-SignedHeaders=host&amp;X-Amz-Signature=1b1ba452ba9238e69bf53e291d73e17e7fc094ed32ddc63e72472a54a74327b1: x509: certificate signed by unknown authority
</code></pre>

**Comment by ericis** [comment]
<p>Correct. <a class="mention" href="/joshbouganim">@joshbouganim</a> makes a great point. GHE 3.x suffers the same error. <a href="https://docs.github.com/en/enterprise-server@3.0/packages/working-with-a-github-packages-registry/working-with-the-docker-registry#downloading-an-image" rel="noopener nofollow ugc">The documentation states</a>:</p>
<blockquote>
<p><strong>Note:</strong> The GitHub Packages Docker registry will be superseded in a future GitHub Enterprise Server release with the Container registry, which offers improved container support.<br>
…<br>
<strong>Note:</strong> You must pull the image using <code>IMAGE_NAME:VERSION</code> and not using <code>IMAGE_NAME:SHA</code> .</p>
</blockquote>
<p>However, while using that tag format and not SHA, the documentation does not indicate that a known, rather alerting docker “warning” (error) will be displayed.</p>

**Comment by fewkz** [comment]
Is pulling images from the registry by digest available yet?

**Reply by fewkz** [reply]
Just tested, looks like pulling by digest is available now

**Reply by asadsubhan0** [reply]
I'm not sure this issue is resolved, as I am trying to pull the image from github packages using SHA digest and its not pulling, whereas when I am pulling the same image from registry other than github packages using SHA digest its pulling fine.

I am using GitHub Enterprise 3.16 version.
